File: src/main/scala/com/xmlcalabash/util/XProcCollectionFinder.scala
package com.xmlcalabash.util

import java.net.URI
import com.xmlcalabash.config.XMLCalabashConfig
import net.sf.saxon.expr.XPathContext
import net.sf.saxon.lib.{CollectionFinder, Resource, ResourceCollection}
import net.sf.saxon.om.{Item, SpaceStrippingRule}
import net.sf.saxon.s9api.XdmNode
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters.IteratorHasAsJava

object XProcCollectionFinder {
  val DEFAULT = "http://xmlcalabash.com/saxon-default-collection"
}

class XProcCollectionFinder(config: XMLCalabashConfig, docs: List[XdmNode], chainedFinder: CollectionFinder)
  extends CollectionFinder {

  protected val logger: Logger = LoggerFactory.getLogger(this.getClass)

  override def findCollection(context: XPathContext, collectionURI: String): ResourceCollection = {
    logger.trace("Collection: " + collectionURI)

    if (collectionURI == null || collectionURI == XProcCollectionFinder.DEFAULT) {
      new DocumentResourceCollection(collectionURI, docs)
    } else {
      val cURI = new URI(collectionURI)
      val docs = config.collection(cURI)
      if (docs.isEmpty) {
        chainedFinder.findCollection(context, collectionURI)
      } else {
        new DocumentResourceCollection(collectionURI, docs)
      }
    }
  }

  private class DocumentResourceCollection(val collectionURI: String, val docs: List[XdmNode]) extends ResourceCollection {
    private val uris = ListBuffer.empty[String]
    private var rsrcs = ListBuffer.empty[Resource]

    override def getCollectionURI: String = collectionURI

    def getResourceURIs(context: XPathContext): java.util.Iterator[String] = {
      if (uris.isEmpty) {
        for (doc <- docs) {
          uris += doc.getBaseURI.toASCIIString
        }
      }
      uris.iterator.asJava
    }

    def getResources(context: XPathContext): java.util.Iterator[Resource] = {
      if (rsrcs.isEmpty) {
        for (doc <- docs) {
          rsrcs += new DocumentResource(doc)
        }
      }
      rsrcs.iterator.asJava
    }

    def isStable(context: XPathContext) = true

    override def stripWhitespace(rules: SpaceStrippingRule): Boolean = false
  }

  private class DocumentResource(val doc: XdmNode) extends Resource {
    override def getResourceURI: String = doc.getBaseURI.toASCIIString

    override def getItem(context: XPathContext): Item = doc.getUnderlyingValue.head

    override def getContentType: String = null
  }


}

File: xmlcalabash/src/main/kotlin/com/xmlcalabash/util/XProcCollectionFinder.kt
package com.xmlcalabash.util

import com.xmlcalabash.documents.XProcDocument
import net.sf.saxon.expr.XPathContext
import net.sf.saxon.lib.CollectionFinder
import net.sf.saxon.lib.Resource
import net.sf.saxon.lib.ResourceCollection
import net.sf.saxon.om.Item
import net.sf.saxon.s9api.XdmNode

class XProcCollectionFinder(val coll: List<XProcDocument>, val chain: CollectionFinder): CollectionFinder {
    companion object {
        val DEFAULT = "https://xmlcalabash.com/default/collection"
    }

    override fun findCollection(context: XPathContext?, collectionUri: String?): ResourceCollection {
        if (collectionUri == DEFAULT) {
            val resources = XProcResourceCollection(coll)
            return resources
        } else {
            return chain.findCollection(context, collectionUri)
        }
    }

    inner class XProcResourceCollection(val coll: List<XProcDocument>): ResourceCollection {
        override fun getCollectionURI(): String {
            return DEFAULT
        }

        override fun getResourceURIs(context: XPathContext?): MutableIterator<String> {
            val uris = mutableListOf<String>()
            for (doc in coll) {
                uris.add(doc.baseURI.toString())
            }
            return uris.iterator()
        }

        override fun getResources(context: XPathContext?): MutableIterator<Resource> {
            val resources = mutableListOf<XProcResource>()
            for (doc in coll) {
                resources.add(XProcResource(doc))
            }
            return resources.iterator()
        }

        override fun isStable(context: XPathContext?): Boolean {
            return true
        }
    }

    inner class XProcResource(val doc: XProcDocument): Resource {
        override fun getResourceURI(): String {
            return doc.baseURI.toString()
        }

        override fun getItem(): Item {
            val item = doc.value
            when (item) {
                is XdmNode -> return item.underlyingValue
                else -> return item as Item
            }
        }

        override fun getContentType(): String {
            return doc.contentType.toString()
        }
    }
}